mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala

```diff
@@ -934,31 +934,56 @@ class LogisticRegression @Since("1.2.0") (
       instances: RDD[Instance],
       actualBlockSizeInMB: Double,
       featuresStd: Array[Double],
+      featuresMean: Array[Double],
       numClasses: Int,
-      initialCoefWithInterceptMatrix: Matrix,
+      initialCoefWithInterceptArray: Array[Double],
       regularization: Option[L2Regularization],
       optimizer: FirstOrderMinimizer[BDV[Double], DiffFunction[BDV[Double]]]) = {
     val numFeatures = featuresStd.length
     val bcFeaturesStd = instances.context.broadcast(featuresStd)
+    val bcFeaturesMean = instances.context.broadcast(featuresMean)
 
-    val standardized = instances.mapPartitions { iter =>
+    val scaled = instances.mapPartitions { iter =>
       val inverseStd = bcFeaturesStd.value.map { std => if (std != 0) 1.0 / std else 0.0 }
       val func = StandardScalerModel.getTransformFunc(Array.empty, inverseStd, false, true)
       iter.map { case Instance(label, weight, vec) => Instance(label, weight, func(vec)) }
     }
 
     val maxMemUsage = (actualBlockSizeInMB * 1024L * 1024L).ceil.toLong
-    val blocks = InstanceBlock.blokifyWithMaxMemUsage(standardized, maxMemUsage)
+    val blocks = InstanceBlock.blokifyWithMaxMemUsage(scaled, maxMemUsage)
       .persist(StorageLevel.MEMORY_AND_DISK)
       .setName(s"training blocks (blockSizeInMB=$actualBlockSizeInMB)")
 
-    val getAggregatorFunc = new BlockLogisticAggregator(numFeatures, numClasses, $(fitIntercept),
-      checkMultinomial(numClasses))(_)
+    val multinomial = checkMultinomial(numClasses)
+    val fitWithMean = !multinomial && $(fitIntercept) &&
+      (!isSet(lowerBoundsOnIntercepts) || $(lowerBoundsOnIntercepts)(0).isNegInfinity) &&
+      (!isSet(upperBoundsOnIntercepts) || $(upperBoundsOnIntercepts)(0).isPosInfinity)
+
+    val costFun = if (multinomial) {
+      // TODO: create a separate BlockMultinomialLogisticAggregator for clearness
+      val getAggregatorFunc = new BlockLogisticAggregator(numFeatures, numClasses,
+         $(fitIntercept), true)(_)
+      new RDDLossFunction(blocks, getAggregatorFunc, regularization, $(aggregationDepth))
+    } else {
+       val getAggregatorFunc = new BlockBinaryLogisticAggregator(bcFeaturesStd, bcFeaturesMean,
+         $(fitIntercept), fitWithMean)(_)
+      new RDDLossFunction(blocks, getAggregatorFunc, regularization, $(aggregationDepth))
+    }
+
+    if (fitWithMean) {
+      var i = 0
+      var adapt = 0.0
+      while (i < numFeatures) {
+        if (featuresStd(i) != 0) {
+          adapt += initialCoefWithInterceptArray(i) * featuresMean(i) / featuresStd(i)
```

Comment:
Probably a naive question but why multiply by the mean rather than subtract? I understand it's to compute an adjustment to the intercept, just wondering where this comes from.

Reply:
the orignial `initialCoefWithInterceptArray = (init_w1,init_w2,...,init_wn,init_intercept)`  is an intial solution for problem:
`y = f(w1*x1/std_x1, w2*x2/std_x2, ..., wn*xn/std_xn, intercept) `;

Here, I need another solution for problem:
`y = f(w1*(x1-avg_x1)/std_x1, w2*(x2-avg_x2)/std_x2, ..., wn*(xn-avg_xn)/std_xn, intercept2) `;

So adjust `init_intercept2 = init_intercept + init_w1*avg_x1/std_x1 + init_w2*avg_x2/std_x2 + ... + init_wn*avg_xn/std_xn`

Reply:
I will add more comments for this adjustment.

Reply:
Ah right I get it. Yeah worth a comment.

Reply:
Oops, I wrote the explanation last night, but didnt submit the review till this morning.